supabase/functions/gdk_stats/index.ts

```diff
@@ -0,0 +1,150 @@
+	{ speciesName: "BIRKE", percentage: 3.610419665283775 },
+	{ speciesName: "HAINBUCHE", percentage: 3.4514717918324726 },
+	{ speciesName: "PLATANE", percentage: 3.3499844777467333 },
+	{ speciesName: "PAPPEL", percentage: 2.8882679987582197 },
+	{ speciesName: "ESCHE", percentage: 2.7732339909124263 },
+	{ speciesName: "KIEFER", percentage: 2.4801738492365874 },
+	{ speciesName: "ULME", percentage: 1.946998560663788 },
+	{ speciesName: "BUCHE", percentage: 1.7521519487483419 },
+	{ speciesName: "HASEL", percentage: 1.1728050122766912 },
+	{ speciesName: "WEIßDORN", percentage: 1.1243755820844975 },
+	{ speciesName: "WEIDE", percentage: 1.0893799565376909 },
+	{ speciesName: "MEHLBEERE", percentage: 0.90469336494228544013 },
+	{ speciesName: "ERLE", percentage: 0.80907628481923630514 },
+	{ speciesName: "APFEL", percentage: 0.70092851296813704739 },
+];
+
+const supabaseServiceRoleClient = createClient(
+	SUPABASE_URL,
+	SUPABASE_SERVICE_ROLE_KEY
+);
+
+const getCount = async (tableName: string): Promise<number> => {
+	const { count } = await supabaseServiceRoleClient
+		.from(tableName)
+		.select("*", { count: "exact", head: true });
+	return count || 0;
+};
+
+const getWateringsCount = async (): Promise<number> => {
+	const beginningOfYear = new Date(`${new Date().getFullYear()}-01-01`);
+	const { count } = await supabaseServiceRoleClient
+		.from("trees_watered")
+		.select("*", { count: "exact", head: true })
+		.gt("timestamp", beginningOfYear.toISOString());
+	return count || 0;
+};
+
+const getPumpsCount = async (): Promise<number> => {
+	const response = await fetch(PUMPS_URL);
+	const geojson = await response.json();
+	return geojson.features.length;
+};
+
+const getMonthlyWaterings = async (): Promise<Monthly[]> => {
+	const { data, error } = await supabaseServiceRoleClient
+		.rpc("calculate_avg_waterings_per_month")
+		.select("*");
+	return data.map((month: any) => ({
+		month: month.month,
+		wateringCount: month.watering_count,
+		averageAmountPerWatering: month.avg_amount_per_watering,
+	}));
+};
+
+const handler = async (request: Request): Promise<Response> => {
+	if (request.method === "OPTIONS") {
+		return new Response(null, { headers: corsHeaders, status: 204 });
+	}
+
+	try {
+		const [
+			usersCount,
+			wateringsCount,
+			adoptionsCount,
+			numPumps,
+			monthlyWaterings,
+		] = await Promise.all([
+			getCount("profiles"),
+			getWateringsCount(),
+			getCount("trees_adopted"),
+			getPumpsCount(),
+			getMonthlyWaterings(),
+		]);
+
+		const stats: GdkStats = {
+			numTrees: TREE_COUNT,
+			numPumps: numPumps,
+			numActiveUsers: usersCount,
+			numWateringsThisYear: wateringsCount,
+			monthlyWaterings: monthlyWaterings,
+			numTreeAdoptions: adoptionsCount,
+			mostFrequentTreeSpecies: MOST_FREQUENT_TREE_SPECIES,
+		};
+
+		return new Response(JSON.stringify(stats), {
+			status: 200,
+			headers: {
+				...corsHeaders,
+				"Content-Type": "application/json",
+			},
+		});
+	} catch (error) {
```

Comment:
Since we are doing several fetch requests it would be nice to have a more sophisticated error handling then just returning 500 and not logging any errors. In Parla we introduced IMO a got pattern for error handling https://github.com/technologiestiftung/parla-api/blob/main/src/lib/errors.ts

Why not adopt this as a standard

Reply:
> and not logging any errors

In fact, I am logging the error at a central point in line 189.
```
} catch (error) {
    console.log(error)
```


I don't see big relevance for having more sophisticated error handling here. I return the expected 200 response if everything is alright or in case of errors, I `console.log(error)` the error for debugging purposes (it will show up in the Supabase Edge functions log) and deliver it with 500 response to the frontend.

We will not display technical backend error messages in the frontend for the data vis feature, rather we display a default error message. Either fetching data for the data vis is successful or it fails, then we display the generic error in the frontend.

What is the advantage of more sophisticated error handling if the frontend does not consume it and act upon it? Or do you suggest that the frontend should handle errors for each of the data vis sections separately?

Reply:
I would have used a more distinguishable logging just to know where the error is coming from. For easier bug tracing. 
Also as I understand it:

.debug(), .info() and .log() send to 1 (stdout) while .warn() and .error() send to 2 (stderr)

Which might be handled different in the logging of Supabase.

Reply:
I implemented more distinguishable error messages in https://github.com/technologiestiftung/giessdenkiez-de-postgres-api/pull/269/commits/fe11dba4c20a5bc8d5d2e5fa36c314ef5add418e
Using Supabase Edge Function Logs, it should be clear enough now what is going on based on the logs provided by console.error()